src/main/java/org/jenkinsci/plugins/gitclient/JGitAPIImpl.java

```diff
@@ -166,15 +179,99 @@ public class JGitAPIImpl extends LegacyCompatibleGitAPIImpl {
 
         // to avoid rogue plugins from clobbering what we use, always
         // make a point of overwriting it with ours.
-        SshSessionFactory.setInstance(new TrileadSessionFactory(hostKeyFactory, listener));
-
+        SshSessionFactory.setInstance(
+                buildSshdSessionFactory(hostKeyFactory.forJGit(listener), asSmartCredentialsProvider()));
         if (httpConnectionFactory != null) {
             httpConnectionFactory.setCredentialsProvider(asSmartCredentialsProvider());
             // allow override of HttpConnectionFactory to avoid JENKINS-37934
             HttpTransport.setConnectionFactory(httpConnectionFactory);
         }
     }
 
+    private SshdSessionFactory buildSshdSessionFactory(
+            AbstractJGitHostKeyVerifier abstractJGitHostKeyVerifier, SmartCredentialsProvider credentialsProvider) {
+
+        if (Files.notExists(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.toPath())) {
+            try {
+                Files.createDirectories(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE
+                        .getParentFile()
+                        .toPath());
+                Files.createFile(SshHostKeyVerificationStrategy.JGIT_KNOWN_HOSTS_FILE.toPath());
+            } catch (IOException e) {
+                LOGGER.log(Level.SEVERE, "cannot create knowhosts file", e);
+            }
+        }
+        return new SshdSessionFactory() {
```

Comment:
the factory will have a reference to the outer object.   does this cause issues if used on an agent?

Reply:
nope